sympy/controlsys/lti.py

```diff
@@ -0,0 +1,1363 @@
+StateSpaceModel(
+Matrix([
+[  0,   1,   0],
+[  0,   0,   1],
+[-a0, -a1, -a2]]),
+Matrix([
+[0],
+[0],
+[1]]),
+Matrix([[b0, b1, b2]]),
+Matrix([[0]]))
+>>> ssm_series = ssm1.cascade(ssm2)
+>>> ssm_series
+StateSpaceModel(
+Matrix([
+[  0,   1,   0,   0,   0],
+[-a0, -a1,   0,   0,   0],
+[  0,   0,   0,   1,   0],
+[  0,   0,   0,   0,   1],
+[ b0,  b1, -a0, -a1, -a2]]),
+Matrix([
+[0],
+[1],
+[0],
+[0],
+[0]]),
+Matrix([[0, 0, b0, b1, b2]]),
+Matrix([[0]]))
+>>> ssm_parallel = ssm2.parallel(ssm2)
+>>> ssm_parallel
+StateSpaceModel(
+Matrix([
+[  0,   1,   0,   0,   0,   0],
+[  0,   0,   1,   0,   0,   0],
+[-a0, -a1, -a2,   0,   0,   0],
+[  0,   0,   0,   0,   1,   0],
+[  0,   0,   0,   0,   0,   1],
+[  0,   0,   0, -a0, -a1, -a2]]),
+Matrix([
+[0],
+[0],
+[1],
+[0],
+[0],
+[1]]),
+Matrix([[b0, b1, b2, b0, b1, b2]]),
+Matrix([[0]]))
+
+>>> var('a, b, s')
+(a, b, s)
+>>> tfm1 = TransferFunctionModel(Matrix([1 / (a * s)]))
+>>> tfm1
+TransferFunctionModel(Matrix([[1/(a*s)]]))
+>>> tfm2 = TransferFunctionModel(Matrix([(b + s) / (a + s)]))
+>>> tfm2
+TransferFunctionModel(Matrix([[(b + s)/(a + s)]]))
+>>> tfm_series = tfm1.cascade(tfm2)
+>>> tfm_series
+TransferFunctionModel(Matrix([[(b + s)/(a*s*(a + s))]]))
+>>> tfm_parallel = tfm1.cascade(tfm2)
+>>> tfm_parallel
+TransferFunctionModel(Matrix([[(b + s)/(a*s*(a + s))]]))
+
+Note that the systems must have a matching nubmer of inputs and outputs for this to work!
+
+Evaluating models
+=================
+
+Models are always evaluated using the `evaluate` method.
+It is done numercialy or symbolicaly, depending on the parameters of the call:
+
+>>> from sympy import Heaviside
+>>> var('t') # symbolic evaluation
+>>> u = Matrix([Heaviside(t)])
+>>> x0 = zeros(3, 1)
+>>> t0 = 0
+>>> y = ssm.evaluate(u, x0, t, t0)
+>>> y
+Matrix([
+[     t],
+[sin(t)]])
+
+>>> y_list = ssm.evaluate(u, x0, (t, [0, 10, 0.1]))
+
+
+Note that the only difference between the calls is that in the second one, t is substituted by `(t, np.arange(0, 10, 0.1))`; a tuple of the symbol t and a list of times to evaluate for.
+The first call will return a matrix valued expression in terms of t.
+The second call will return a list of matrix valued outputs, matching the time input.
+
+For details on syntax and additional methods, see the docstings of each class and its methods.
+
+"""
+from __future__ import division
+
+from sympy import (
+    Symbol, var, simplify, oo, exp,
+    Poly, lcm, LC, degree, Integral, integrate,
+    Matrix, BlockMatrix, eye, zeros,
+    latex, ShapeError, ImmutableMatrix, MutableMatrix,
+    SparseMatrix, MutableDenseMatrix
+)
+
+from sympy.core.compatibility import range
+from sympy.printing import sstr
+
+# import mpmath for numercial results
+from mpmath import expm, quad, matrix as mpm_matrix
+
+__all__ = [
+    'StateSpaceModel',
+    'TransferFunctionModel'
+]
+
+_matrixTypes = (
+    Matrix, ImmutableMatrix, MutableMatrix, SparseMatrix, MutableDenseMatrix)
+
+
+class StateSpaceModel(object):
```

Comment:
These can likely be subclasses from some sympy base objects, no?

Reply:
It's not clear to me why there are classes here at all. Maybe I don't understand this well enough but I imagine that I would have written this module as a few functions.

If a StateSpaceModel makes sense as a symbolic entity then probably it should subclass Basic so that you can do subs etc.

Reply:
I think classes make sense. These are fundamental things in control system theory and they have attributes and you can do numerous operations on them. I think classes make sense. It is exactly what I would have done.

Reply:
In formal terms how would you define these entities? It seems to me that a "state space model" is a way of representing a system as ODEs and a transfer function is a property of that system.

If I was trying to solve a practical problem in this domain would it make sense that I would start out with the matrices or would I be starting from ODEs and then trying to get them in this form?

Taking ODEs as the starting point in the API rather than matrices makes it possible to extend this for nonlinear systems later.

Reply:
Yes, they are both representations of linear ODEs.

> If I was trying to solve a practical problem in this domain would it make sense that I would start out with the matrices or would I be starting from ODEs and then trying to get them in this form?

You typically develop a model of a physical system and arrive at a set of linear ODEs, then you put them in state space matrix or transfer function form. After that you use those forms and to find the solutions to the ODEs and various other properties.

Reply:
You can see how people work with these things here:

https://github.com/python-control/python-control

That's a numeric package that does the kinds of things this package does. You can also see the matlab docs for similar ideas:

https://www.mathworks.com/help/control/ref/ss.html

In these programs you start by creating state space matrices of ratios of polynomials for transfer functions. This PR introduces a similar idea and starting point.